docs/madr/decisions/056-meshservice-migration.md

```diff
@@ -0,0 +1,349 @@
+
+### Generate MeshMultiZoneService
+
+Generate `MeshMultiZoneService` and generate hostnames equal to the
+existing `<name>_<namespace>_svc_<port>` format. Users need to disable this
+autogeneration to get the local zone behavior.
+
+Policy matching would need to be adjusted so that old policies also target
+`MeshMultiZoneService`.
+
+#### Migration
+
+This migration is potentially the trickiest.
+
+1. Upgrade to new version
+1. `MeshMultiZoneService` are now generated and they take over generation of
+   `name_namespace_svc_port` clusters. `MeshServices` are also generated
+
+It's not clear what the path or even the goal is from here.
+
+#### Pros
+
+- Makes explicit the fact that requests are going cross-zone
+- Things look exactly the same as how we want cross-zone services to work
+
+#### Cons
+
+- Ihis would require all zones to be upgraded so that MeshServices are synced
+  and can be matched by MeshMultiZoneService
+- Otherwise the most complex option
+
+### Generate new clusters
+
+Instead of `MeshService` taking over cluster generation, generate separate clusters so that:
+
+- `<kuma.io/service>.mesh` -> Old Kuma VIP -> old cluster
+- `HostnameGenerator` DNS name for the `MeshService` -> New Kuma VIP
+  - this includes synced `MeshServices`
+- `name.namespace.svc.cluster.local` -> ClusterIP -> depends
+
+The new format is:
+
+```
+<name>_<namespace>_<zone>_<mesh>_msvc_<port>
+```
+
+where `<namespace>` is empty for universal.
+
+This means increasing the number of synced clusters when **not using reachable
+services**, during the migration period where both kinds of VIPs are in use and
+thus both clusters are required.
+
+However, with reachable services, things become very straightforward because it's always
+very explicit which services are actually in use and need to be synced.
+
+#### Kubernetes DNS ClusterIP
+
+Kubernetes has an extra complication because the Kube DNS name/ClusterIP
+has to point at one cluster or the other.
+
+We either need a switch:
+
+```
+KUMA_RUNTIME_KUBERNETES_DNS_CLUSTER_PRIORITY=ServiceTag|MeshService
+```
+
+or maybe it's OK to switch to local-zone load balancing:
+
+- we currently only add an outbound for a service on k8s if there's a local
+  instance of it
+- with `localityAwareLoadBalancing: false`
+  - only if local instances becomes unhealthy does traffic flow to other zones
+- with `localityAwareLoadBalancing: true`
+  - traffic never flows cross-zone
+- what about `MeshLoadBalancingStrategy`?
+  - only `localZone` applies
+
+#### Migration
+
+1. Upgrade to new version
+  If the user doesn't change any config, nothing changes. No `MeshServices` are generated.
+1. On `Mesh`, users have the option `meshServices`:
+  - `Disabled` is the default
+  - `Everywhere` syncs `MeshService` outbounds to all data plane proxies
+  - `Exclusive` disables `kuma.io/service` outbounds and enables `MeshService`
+  - `ReachableBackendRefs` to generate all `MeshServices` but sync
```

Comment:
Does it disable kuma.io/service as well?

Reply:
No. The idea being you'd still choose what to sync via `reachable*`